contracts/GuildPin.sol

```diff
@@ -0,0 +1,249 @@
+
+    /// @notice Empty space reserved for future updates.
+    uint256[42] private __gap;
+
+    /// @notice Sets metadata and the associated addresses.
+    /// @param name The name of the token.
+    /// @param symbol The symbol of the token.
+    /// @param treasury The address where the collected fees will be sent.
+    /// @param _validSigner The address that should sign the parameters for certain functions.
+    function initialize(
+        string memory name,
+        string memory symbol,
+        address payable treasury,
+        address _validSigner
+    ) public initializer {
+        validSigner = _validSigner;
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+        __SoulboundERC721_init(name, symbol);
+        __TreasuryManager_init(treasury);
+    }
+
+    // Note: the reInitialize function was last used with version 2
+    // function reInitialize() public reinitializer(2) {
+    // }
+
+    function claim(
+        PinDataParams memory pinData,
+        address payable adminTreasury,
+        uint256 adminFee,
+        uint256 signedAt,
+        string calldata cid,
+        bytes calldata signature
+    ) external payable {
+        if (signedAt < block.timestamp - SIGNATURE_VALIDITY) revert ExpiredSignature();
+        if (
+            claimedTokens[pinData.receiver][pinData.guildAction][pinData.guildId] != 0 ||
+            claimerUserIds[pinData.userId][pinData.guildAction][pinData.guildId]
+        ) revert AlreadyClaimed();
+        if (!isValidSignature(pinData, adminTreasury, adminFee, signedAt, cid, signature)) revert IncorrectSignature();
+
+        uint256 tokenId = totalSupply() + 1;
+
+        unchecked {
+            ++totalMintedPerGuild[pinData.guildId];
+        }
+
+        claimedTokens[pinData.receiver][pinData.guildAction][pinData.guildId] = tokenId;
+        claimedTokensDetails[tokenId] = PinData(
+            pinData.receiver,
+            pinData.guildAction,
+            uint88(pinData.userId),
+            pinData.guildName,
+            uint128(pinData.guildId),
+            uint128(totalMintedPerGuild[pinData.guildId]),
+            uint128(block.timestamp),
+            uint128(pinData.createdAt)
+        );
+        cids[tokenId] = cid;
+        claimerUserIds[pinData.userId][pinData.guildAction][pinData.guildId] = true;
+
+        // Fee collection
+        uint256 guildFee = fee;
+        if (msg.value == guildFee + adminFee) {
+            treasury.sendEther(guildFee);
+            if (adminTreasury != address(0)) adminTreasury.sendEther(adminFee);
+        } else revert IncorrectFee(msg.value, guildFee + adminFee);
+
+        _safeMint(pinData.receiver, tokenId);
+
+        emit Locked(tokenId);
+
+        emit Claimed(pinData.receiver, pinData.guildAction, pinData.guildId);
+    }
+
+    function burn(uint256 userId, GuildAction guildAction, uint256 guildId) external {
+        uint256 tokenId = claimedTokens[msg.sender][guildAction][guildId];
```

Comment:
Wrote a test to see better what happens here: 79d5059
I could revert with a custom error here (I already have a `NonExistentToken` error in the codebase, or come up with a new one), however, I don't think that would really be an improvement over the current situation (and not worth the ~8k gas extra).
You might say that since the revert comes from the underlying dependency, it might change just from a dependency update, but I will notice that now thanks to the test. Thanks for bringing this to my attention!

Reply:
Yeah, that makes sense! And for sure!